### DISCUSSION TITLE
Counterfactual Regret Minimization (CFR) without generating all the states apriori

### DISCUSSION THREAD

**Discussion by PURANJAY14** [deepmind/open_spiel#1239] [Q&A] [answered]
Hi,
Is there a way to run counterfactual regret minimization (CFR) without precomputing (get_all_states.py) all the states? The intermediate policies updated during CFR often have sparse action probabilities, making it unnecessary to traverse the entire game tree.

**Comment by lanctot** [comment]
CFR should not require getting all states. 

I just checked [cfr.py](https://github.com/google-deepmind/open_spiel/blob/master/open_spiel/python/algorithms/cfr.py) and I don't see an import for get_all_states.py or where it gets all the states. Can you point to where you think the CFR implementation gets all the states?

Oh wait, do you mean information states? You may want to look into MCCFR in that case.

**Comment by PURANJAY14** [comment]
In the [policy.py](https://github.com/google-deepmind/open_spiel/blob/master/open_spiel/python/policy.py?plain=1#L242) which is accessed inside [cfr.py](https://github.com/google-deepmind/open_spiel/blob/master/open_spiel/python/algorithms/cfr.py?plain=1#L165).

**Comment by lanctot** [comment, marked as answer]
Ah yes, but to be honest I really don't remember why it does it that way. There was a deep technical reason. @jblespiau, do you remember?

@PURANJAY14 if I were you, I would make an alternative tabular policy called TabularPolicyDict that simply wraps a python dictionary but otherwise exposes the necessary methods in the Policy base class. (And initializing the state policy to uniform + adding it to the dictionary upon first call). Then replace all the ones in cfr.py with your custom dict.

Hope this helps!

**Reply by PURANJAY14** [reply]
Thanks for the idea

**Reply by lanctot** [reply]
No problem. If you do it, I would encourage you to submit a PR so others might benefit from it.
